server-timing/defaults.php

```diff
@@ -211,3 +211,85 @@ static function( $acc, $query ) {
 	}
 }
 add_action( 'wp_loaded', 'perflab_register_default_server_timing_template_metrics' );
+
+/**
+ * Registers additional Server-Timing metrics as configured in the setting.
+ *
+ * These metrics should be registered as soon as possible. They can be added
+ * and modified in the "Tools > Server-Timing" screen.
+ *
+ * @since n.e.x.t
+ */
+function perflab_register_additional_server_timing_metrics_from_setting() {
+	$options = (array) get_option( PERFLAB_SERVER_TIMING_SETTING, array() );
+
+	/*
+	 * This closure measures performance of a hook (action or filter), as follows:
+	 *
+	 * 1. Add a hook callback at the minimum (i.e. earliest) priority possible.
+	 * 2. In that callback, register the metric for the hook, with a prefix of either "action" or "filter".
+	 * 3. Provide a measuring callback which captures the time span between beginning to end of the hook:
+	 *     1. Capture the current time immediately, i.e. at the earliest hook priority.
+	 *     2. Add another hook callback at the maximum (i.e. latest) priority possible.
+	 *     3. In that callback, capture the current time, leading the Server-Timing API to calculate the difference.
+	 *
+	 * Parameters to this closure are the hook name, what type of hook it is (either "action" or "filter", used as a
+	 * prefix for the registered metric), and the callback function to add the hook (either "add_action" or
+	 * "add_filter").
+	 */
+	$measure_hook = static function( $hook_name, $hook_type, $add_hook_func ) {
+		$metric_slug      = "{$hook_type}-{$hook_name}";
+		$measure_callback = static function( $metric ) use ( $hook_name, $add_hook_func ) {
+			$metric->measure_before();
+			call_user_func(
+				$add_hook_func,
+				$hook_name,
+				static function( $passthrough = null ) use ( $metric ) {
+					$metric->measure_after();
+					return $passthrough;
+				},
```

Comment:
Doesn't this callback need to remove itself here? Otherwise, if a hook gets called multiple times it seems there will end up being leftover callbacks at `PHP_INT_MAX` that will try to keep calling `measure_after` on the previous `$metric`.

Reply:
Done in https://github.com/WordPress/performance/pull/784/commits/447f078523412aa5184b75f04b28dfe6da9dc5a3